bigquery-connector-common/src/main/java/com/google/cloud/bigquery/connector/common/BigQueryDirectDataWriterHelper.java

```diff
@@ -184,19 +193,53 @@ public void addRow(ByteString message) throws IOException {
     appendRequestRowCount++;
   }
 
+  /**
+   * Thorws an exception if there is an error in any of the responses received thus far. Since
+   * responses arrive in order, we proceed to check the next response only after the previous
+   * response has arrived.
+   *
+   * @throws BigQueryConnectorException If there was an invalid response
+   */
+  private void checkForFailedResponse(boolean waitForResponse) {
+    while (nextIndexCheckResponse < appendRowsFutures.size()) {
+      ApiFuture<AppendRowsResponse> validatedAppendRowsFuture =
+          appendRowsFutures.get(nextIndexCheckResponse);
+      if (waitForResponse || validatedAppendRowsFuture.isDone()) {
+        try {
+          AppendRowsResponse response = validatedAppendRowsFuture.get();
+          // If we got here, the response was successful
+          nextIndexCheckResponse++;
+        } catch (InterruptedException | ExecutionException e) {
+          logger.error(
+              "Write-stream {} with name {} exception while retrieving AppendRowsResponse",
+              partitionId,
+              writeStreamName);
+          throw new BigQueryConnectorException("Exception while retrieving AppendRowsResponse", e);
+        }
```

Comment:
add a finally block here to shutdown the executor?

Reply:
Done